cmd/kubelet/app/server.go

```diff
@@ -440,6 +449,130 @@ func InitializeTLS(s *options.KubeletServer) (*server.TLSOptions, error) {
+	if err != nil {
+		return fmt.Errorf("unable to create API server client config: %v", err)
+	}
+
+	certificatesclient, err := unversionedcertificates.NewForConfig(clientConfig)
+	if err != nil {
+		return fmt.Errorf("unable to create certificates client: %v", err)
+	}
+
+	// (2).
+	hostname := nodeutil.GetHostname(s.HostnameOverride)
+
+	var ips []net.IP
+	nodeIP := net.ParseIP(s.NodeIP)
+	if nodeIP != nil {
+		ips = []net.IP{nodeIP}
+	}
+
+	commonName := fmt.Sprintf("kubelet-%s", hostname)
+	req, err := utilcertificates.NewCertificateRequest(s.TLSPrivateKeyFile, &pkix.Name{CommonName: commonName}, []string{hostname}, ips)
+	if err != nil {
+		return fmt.Errorf("unable to generate certificate request: %v", err)
+	}
+
+	// (3).
+	certificate, err := requestCertificate(certificatesclient, req, 3600) // Make a default timeout = 3600s
+	if err != nil {
+		return fmt.Errorf("unable to request certificate from API server: %v", err)
+	}
+
+	// (4).
+	if err := crypto.WriteCertToPath(s.TLSCertFile, certificate); err != nil {
+		return fmt.Errorf("unable to create certificate file on disk: %v", err)
+	}
+
+	return nil
+}
+
+// requestCertificate sends the certificate signing request to API server and watches the object's status.
+// It returns the API server's issued certificate (pem-encoded) on success.
+// If there is any errors, or the watch timeouts, it returns an error.
+func requestCertificate(client unversionedcertificates.CertificateSigningRequestsGetter, request []byte, defaultTimeoutSeconds int64) (certificate []byte, err error) {
+	if _, err = client.CertificateSigningRequests().Create(&certificates.CertificateSigningRequest{
+		TypeMeta:   unversioned.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: api.ObjectMeta{GenerateName: "csr-"},
+
+		// Username, UID, Groups will be injected by API server.
+		Spec: certificates.CertificateSigningRequestSpec{Request: request},
+	}); err != nil {
+		return nil, fmt.Errorf("cannot create certificate signing request: %v", err)
+
+	}
+
+	resultCh, err := client.CertificateSigningRequests().Watch(api.ListOptions{
+		Watch:          true,
+		TimeoutSeconds: &defaultTimeoutSeconds,
+		// Label and field selector are not used now.
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot watch on the certificate signing request: %v", err)
+	}
+
+	var status certificates.CertificateSigningRequestStatus
+	ch := resultCh.ResultChan()
+
+	for {
+		event, ok := <-ch
+		if !ok {
+			break
+		}
+
+		if event.Type == watch.Modified {
+			status = event.Object.(*certificates.CertificateSigningRequest).Status
+			for _, c := range status.Conditions {
+				if c.Type == certificates.CertificateDenied {
+					return nil, fmt.Errorf("certificate signing request is not approved: %v, %v", c.Reason, c.Message)
+				}
+				if c.Type == certificates.CertificateApproved && status.Certificate != nil {
+					return status.Certificate, nil
+				}
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("watch channel closed")
```

Comment:
on timeout, should we try to clean up our CSR? (the server might deny deleting it, which is fine, but it seems like we should at least try)